[REVIEW] Switch Models to use Crossfit

VibhuJawa opened the pull request:
This PR enables using Crossfit. 

Todo:

- [x]  Domain Model
- [x] Quality Model 
- [x] Update installation instructions  (Verified locally)

Benchmarks: 

| Subset Place                     | Dataset Size | Batch Size           | GPUs   | Time Taken | Implementation | Speedup | Note                                         |
|----------------------------------|--------------|----------------------|--------|------------|----------------|---------|----------------------------------------------|
| subset_CC-MAIN-2023-14_english   | 1.8 GB       | 1024** (dynamic)     | 16 V100| 163        | Crossfit       | 1       | Batch size is dynamic                        |
| subset_CC-MAIN-2023-14_english   | 1.8 GB       | 256 (static)         | 16 V100| 230        | MainLine       | 1.41    | Batch size is static, 512,1024 both OOM      |

![image](https://github.com/NVIDIA/NeMo-Curator/assets/4837571/dac69ea7-f136-4577-9b93-a7b175bb230e)


A100 numbers: 



![image](https://github.com/NVIDIA/NeMo-Curator/assets/4837571/ff1c2316-903a-4460-b8ee-ae9c03361d8e)

| Dataset                         | Size (Mb) | Resolution | GPU    | Time (s) | Model    | Speedup | Batch Size         |
|---------------------------------|-----------|------------|--------|----------|----------|---------|--------------------|
| subset_CC-MAIN-2023-14_english  | 148       | 1024**     | 2 A100 | 50.558   | Crossfit | 1       | Batch size is dynamic |
| subset_CC-MAIN-2023-14_english  | 148       | 1024**     | 2 A100 | 78.487   | Mainline | 1.55    | Batch size is static   |

VibhuJawa commented:
CC: @sarahyurick for an initial review.

ayushdg commented:
Side note: NeMo-Curator requires all PR's to include both signed commits as well as signed-off. This can be enabled with `git commit -sS ...`.
There's some more info in https://github.com/NVIDIA/NeMo-Curator/blob/main/CONTRIBUTING.md#pull-requests-pr-guidelines.

VibhuJawa commented:
> Side note: NeMo-Curator requires all PR's to include both signed commits as well as signed-off. This can be enabled with `git commit -sS ...`. There's some more info in https://github.com/NVIDIA/NeMo-Curator/blob/main/CONTRIBUTING.md#pull-requests-pr-guidelines.

Thanks, fixed it.

VibhuJawa commented on nemo_curator/distributed_data_classification/domain_classifier_inference.py:
```diff
@@ -17,162 +17,130 @@
-    tokenizer = DebertaV2TokenizerFast.from_pretrained(cfg.model)
-    cfg.tokenizer = tokenizer
-    return cfg
+class CustomModel(nn.Module):
```
I think we should just remove this file in favor of API example or edit it to use API directly.  Dont think we should have 2  implementations .

sarahyurick replied:
Yeah, I'm ok with that. The original intention with the `*_classifier_inference.py` scripts was to give the user some insight on how to write a custom MNMG script for their own LLM classifiers (that they had trained themselves). But I think we're moving away from that with the API anyway, and the user can always dig into our other MNMG code if they're really interested.

VibhuJawa replied:
Please take another look now.

sarahyurick commented on nemo_curator/distributed_data_classification/domain_classifier_inference.py:
```diff
@@ -17,162 +17,130 @@
+        config: A config object.
         device: A specified PyTorch device, such as torch.device("cuda") or torch.device("cpu").
         model_path: The path to the model file.
+        autocast: Wether to autocast or not
```
Minor typo here.

VibhuJawa replied:
fixed

sarahyurick reviewed (commented):
I haven't been able to try running this yet (conda environments are acting up so I'm running an extensive cleanup right now) but wanted to leave at least an initial review for now. Changes look good and very straightforward to me! I don't have any major concerns or issues with this; I plan to approve/request further changes as needed later today when I actually run it.

Only question atm is whether you're planning to add the quality classifier in this PR or sometime later? Either way is ok with me.

VibhuJawa commented on examples/distributed_data_classification_examples/domain_api_example.py:
```diff
@@ -53,11 +53,13 @@ def main(args):
     ]
 
-    model_file_name = "/path/to/pytorch_model_file.pth"
+    model_file_name = "/home/nfs/syurick/LLM_domain_classifier_inference/GoogleDebertaAgree_v3b_bce_maxlen512_bs64_noRef_best.pth"
```
Yes, thats correct

sarahyurick commented on examples/distributed_data_classification_examples/domain_api_example.py:
```diff
@@ -68,7 +70,7 @@ def main(args):
         model_file_name=model_file_name,
         labels=labels,
-        filter_by=["Games", "Sports"],
+        # filter_by=["Games", "Sports"],
```
LMK if I can help debug this. Are you planning to fix it in this PR or later?

VibhuJawa replied:
Will do, trying to fix it in this PR.

VibhuJawa replied:
Done

sarahyurick reviewed (commented):
I was able to run this as is! Thanks again.

VibhuJawa commented on nemo_curator/distributed_data_classification/domain_classifier_inference.py:
```diff
@@ -238,20 +205,22 @@ def main():
-            enforce_metadata=False,
+        df["sliced_text"] = df["text"].str.slice(0, max_chars)
+        columns_to_keep_list = df.columns.to_list()
+        columns_to_keep_list.remove("sliced_text")
```
Thanks

VibhuJawa commented on nemo_curator/modules/distributed_data_classifier.py:
```diff
@@ -13,56 +13,131 @@
+    max_len = 512
+
+
+class CustomModel(nn.Module):
```
So i thought about it,  i think for now, we should keep it here in favor of less abstraction .  Like for a new user its easier if all of this code in a single place . If this file becomes too big, we can move it out .

VibhuJawa commented on examples/distributed_data_classification_examples/test_models.ipynb:
```diff
@@ -0,0 +1,373 @@
+{
```
I can move this file if we don't like having notebooks but personally i think for examples notebooks are a better way to understand.

ryantwolf replied:
Actually while we're moving stuff around, do you mind also moving everything in `distributed_data_classification_examples` to the root `examples` folder? I think they should be with everything else. And rename the files too please.
- `domain_api_example.py` -> `domain_classifier_example.py`
- `quality_api_example.py` -> `quality_classifier_example.py`

VibhuJawa commented:
@sarahyurick , I have added the quality model and cleaned up scripts, Can i get a re-review please.

ryantwolf commented on nemo_curator/modules/distributed_data_classifier.py:
```diff
@@ -96,6 +150,126 @@ def _filter_documents(
+            autocast=self.autocast,
+        )
+        model = model.to(device)
+        if os.path.exists(self.model_path):
```
Given that the models are going up on HuggingFace soon too, would it make sense to add an additional check after the local disk check to see if the `model_path` is valid on HF and if so get it from there?

VibhuJawa replied:
Lets do that in a followup .

VibhuJawa replied:
To track: https://github.com/NVIDIA/NeMo-Curator/issues/72

ryantwolf commented on nemo_curator/modules/distributed_data_classifier.py:
```diff
@@ -96,6 +150,126 @@ def _filter_documents(
+            if version.parse(TRANSFORMERS_VERSION) >= version.parse("4.31.0"):
+                sd.pop("model.embeddings.position_ids", None)
+            model.load_state_dict(sd, strict=True)
+        return model.eval()
```
If `model_path` doesn't exist, does just the base model get returned? We should probably throw an error instead.

VibhuJawa replied:
Resolved

ryantwolf commented on examples/distributed_data_classification_examples/test_models.ipynb:
```diff
@@ -0,0 +1,373 @@
+    "    )\n",
+    "elif classifier_type == \"QualityClassifier\":\n",
+    "    quality_labels = [\"High\", \"Medium\", \"Low\"]\n",
+    "    model_file_name = \"/home/nfs/syurick/LLM_quality_classifier_inference/\" + \\\n",
```
This path should be changed to something more generic like `quality_classifier.pth`

VibhuJawa replied:
Done

ryantwolf commented on nemo_curator/modules/distributed_data_classifier.py:
```diff
@@ -96,6 +150,126 @@ def _filter_documents(
+    max_chars: int,
+    batch_size: int,
+    label_col: str,
+    prob_col: str = None,
```
Instead of having two variables `prob_col` and `keep_prob`, could you have just one variable `prob_col` and just set `keep_prob = prob_col is not None`?

VibhuJawa replied:
Resolved

ryantwolf commented on examples/distributed_data_classification_examples/test_models.ipynb:
```diff
@@ -0,0 +1,373 @@
+    "%%time\n",
+    "\n",
+    "result_dataset = classifier(dataset=input_dataset)\n",
+    "result_dataset.df = result_dataset.df.rename(columns={\"labels\": f\"{classifier_type}_prediction\"})\n",
```
Doesn't the `"labels"` column get renamed automatically to `pred_column` set in the constructor of the classifier?

VibhuJawa replied:
Removed, was left by mistake

ryantwolf commented on examples/distributed_data_classification_examples/test_models.ipynb:
```diff
@@ -0,0 +1,373 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Run the actuall Classifier"
```
```suggestion
    "# Run the classifier"
```

ryantwolf commented on examples/distributed_data_classification_examples/test_models.ipynb:
```diff
@@ -0,0 +1,373 @@
+{
+ "cells": [
```
It would be good if you could add a text cell at the beginning describing what the goal of the notebook is. Just a simple paragraph describing the inputs, outputs, and requirements.

VibhuJawa replied:
Done

ryantwolf commented on examples/distributed_data_classification_examples/test_models.ipynb:
```diff
@@ -0,0 +1,373 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Verify The file was written correctly"
```
```suggestion
    "#### Inspect the output"
```

VibhuJawa replied:
Resolved

ryantwolf commented on examples/distributed_data_classification_examples/test_models.ipynb:
```diff
@@ -0,0 +1,373 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "##### cleanup the output file"
```
```suggestion
    "##### Cleanup the output file"
```

VibhuJawa replied:
Resolved

ryantwolf commented on examples/distributed_data_classification_examples/test_models.ipynb:
```diff
@@ -0,0 +1,373 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Data File Paths "
```
```suggestion
    "# Define the data file paths "
```

VibhuJawa replied:
Resolved

ryantwolf commented on examples/distributed_data_classification_examples/test_models.ipynb:
```diff
@@ -0,0 +1,373 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Silence Warnings (Majorly HF internal warnings)"
```
```suggestion
    "#### Silence HuggingFace internal warnings)"
```

sarahyurick replied:
There's still an extra parenthesis here.

VibhuJawa replied:
Resolved

ryantwolf reviewed (changes requested):
I'm super glad to see a speedup like this, and thanks so much for removing a lot of code. I left a lot of nitpicks, and a couple of additional things I'd like to see cleaned up. Thanks for doing this!

VibhuJawa commented on setup.py:
```diff
@@ -60,6 +60,7 @@
         "presidio-anonymizer==2.2.351",
         "usaddress==0.5.10",
         "nemo_toolkit[nlp]>=1.23.0",
+        "crossfit @ git+https://github.com/rapidsai/crossfit.git@1ee3de4",
```
Its still a WIP to upload it, still figuring out packaging for the code ,  happy to discuss more if we think we can depend on github repo directly.

VibhuJawa commented on setup.py:
```diff
@@ -68,6 +69,7 @@
         "cuda12x": [
             "cudf-cu12>=24.2",
             "dask-cudf-cu12>=24.2",
+            "cuml-cu12>=24.2",
```
So current behavior of crossfit is to fall back to CPU for machine learning if cuML is not present .  We need to have everything end to end on GPU. 

This will be resolved once we package crossfit up. Lets discuss if we think not having a package for crossfit is a problem. I am trying to get that resolved but did not want that to block adoption here.

VibhuJawa commented on nemo_curator/modules/distributed_data_classifier.py:
```diff
@@ -13,56 +13,117 @@
         self,
-        model_file_name,
+        model,
         labels,
```
Lets do this in a followup PR when we make the HF release changes. Will file an issue to track this.

VibhuJawa replied:
Added issue to track this: https://github.com/NVIDIA/NeMo-Curator/issues/71

VibhuJawa commented on nemo_curator/modules/distributed_data_classifier.py:
```diff
@@ -96,6 +150,126 @@ def _filter_documents(
+    batch_size: int,
+    label_col: str,
+    prob_col: str = None,
+    keep_prob: bool = False,
```
Added

VibhuJawa commented on nemo_curator/modules/distributed_data_classifier.py:
```diff
@@ -13,56 +13,117 @@
+
+
+@dataclass
+class quality_Config:
```
Added

VibhuJawa commented on nemo_curator/modules/distributed_data_classifier.py:
```diff
@@ -13,56 +13,117 @@
+
+
+@dataclass
+class domain_Config:
```
Added

VibhuJawa commented on examples/distributed_data_classification_examples/test_models.ipynb:
```diff
@@ -0,0 +1,373 @@
+    "    \"Sports\",\n",
+    "    \"Travel_and_Transportation\",\n",
+    "    ]\n",
+    "    model_file_name = \"/home/nfs/syurick/LLM_domain_classifier_inference/\" + \\\n",
```
Done

VibhuJawa commented on nemo_curator/distributed_data_classification/quality_classifier_inference.py:
Done

VibhuJawa commented on nemo_curator/distributed_data_classification/domain_classifier_inference.py:
Done

sarahyurick commented on examples/quality_classifier_example.py:
At https://github.com/NVIDIA/NeMo-Curator/blob/main/examples/distributed_data_classification_examples/quality_api_example.py#L36 can you change it to `input_dataset = DocumentDataset.read_json(` (instead of `from_json`)? Thanks!

VibhuJawa replied:
Done, thanks for the close review

sarahyurick commented on nemo_curator/distributed_data_classification/quality_classifier_multiple_models_inference.py:
Do we still want a multiple models example to exist somewhere?

For now maybe we can either open an issue for it or add a TODO in `distributed_data_classifier.py`?

VibhuJawa replied:
Can you open an issue for it, dont think we need it right now but if we do we can add it later on.

sarahyurick replied:
Opened here: https://github.com/NVIDIA/NeMo-Curator/issues/70

VibhuJawa commented:
Really sorry for the commit noise guys (@ryantwolf , @sarahyurick ), i messed up one of the git rebase (forgot to sign off a commit). 

This should be ready for another review now.  Have addressed all the reviews and added issues to follow up ones that i could not get to in this PR.

ryantwolf reviewed (commented):
Great, I missed one file in my initial review, but other than that mostly just nits. Thanks!

VibhuJawa commented on nemo_curator/scripts/classifier_arg_utils.py:
Resolved via https://github.com/NVIDIA/NeMo-Curator/pull/58/commits/ca713c8c03c3242877972df6d61598c3660513d5

VibhuJawa replied:
Resolved

VibhuJawa commented on nemo_curator/modules/distributed_data_classifier.py:
```diff
@@ -96,238 +197,261 @@ def _filter_documents(
+                out_dim = len(labels)  # Multiclass classification
 
         self.prob_column = prob_column
         self.max_len = max_len
```
That was left due to a bad merge. Removed it.

VibhuJawa replied:
https://github.com/NVIDIA/NeMo-Curator/pull/58/commits/ca713c8c03c3242877972df6d61598c3660513d5#diff-e8b5553689ebf070c6bf1a1b8e1d35d771c7cce1f0e042ea8bb9c990783d39ae

VibhuJawa commented on nemo_curator/scripts/domain_classifier_inference.py:
Resolved via https://github.com/NVIDIA/NeMo-Curator/pull/58/commits/1f39839a22fa1e94737fd07fc615a6e27e2d568b

VibhuJawa commented on nemo_curator/modules/distributed_data_classifier.py:
```diff
@@ -13,56 +13,164 @@
+            return self._forward(batch)
+
+
+class CustomModel(nn.Module):
```
That was left due to a bad merge, removed it now. 
https://github.com/NVIDIA/NeMo-Curator/pull/58/commits/ca713c8c03c3242877972df6d61598c3660513d5#diff-e8b5553689ebf070c6bf1a1b8e1d35d771c7cce1f0e042ea8bb9c990783d39ae

VibhuJawa commented:
@ryantwolf , Thanks again for all the careful reviews. Appreciate the help . I think i should have those resolved.

ryantwolf commented on nemo_curator/utils/script_utils.py:
```diff
@@ -165,6 +165,75 @@ def parse_gpu_dedup_args(
+        required=True,
+    )
+    parser.add_argument(
+        "--input-file-type",
```
I made another comment but I'll leave this here as well that we should add a similar arg for `--output-file-type`, you can default it to be the same as `--input-file-type`.

VibhuJawa replied:
Sorry for missing that earlier, resolved via : https://github.com/NVIDIA/NeMo-Curator/pull/58/commits/c6b5d9ed03dc790ab9b675b19de015478c686746

ryantwolf commented on setup.py:
```diff
@@ -98,10 +100,9 @@
             "gpu_exact_dups=nemo_curator.scripts.find_exact_duplicates:console_script",
             "deidentify=nemo_curator.scripts.find_pii_and_deidentify:console_script",
             "generate_statistics=nemo_curator.distributed_data_classification.generate_statistics:console_script",
```
This needs to be removed since `generate_statistics` is gone now, right?

VibhuJawa replied:
Resolved via https://github.com/NVIDIA/NeMo-Curator/pull/58/commits/c6b5d9ed03dc790ab9b675b19de015478c686746

ryantwolf reviewed (commented):
Looking good! Just a couple of nits and things I think you missed from my last review. I'll run some tests to do some final verifications too.

ryantwolf reviewed (approved):
Good with me, thanks!

sarahyurick commented on nemo_curator/distributed_data_classification/generate_statistics.py:
Just wanted to make sure there's nothing here we want to keep?

VibhuJawa replied:
We can add it back as a followup if we think we need it .

sarahyurick commented on tutorials/distributed_data_classification/distributed_data_classification.ipynb:
```diff
@@ -0,0 +1,376 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "input_file_path=\"/input_data_dir/\"\n",
+    "output_file_path = \"output_data_dir/\"\n",
+    "domain_model_path = \"domain_model.pth\"\n",
+    "quality_model_path = \"quality_model.pth\""
```
I see you opened https://github.com/NVIDIA/NeMo-Curator/issues/72, I think once we add support for that we should use real links, paths, etc. so that the user can run it without changing anything.

Should we open an issue for this?

VibhuJawa replied:
Lets add to the issue for that to track we also change this file. Thanks for the suggestion. 👍🏼

sarahyurick reviewed (approved):
LGTM, thanks! Added a couple questions for general discussion.
